crates/runtime/src/memfd.rs

```diff
@@ -424,34 +460,49 @@ impl MemFdSlot {
     #[allow(dead_code)] // ignore warnings as this is only used in some cfgs
     pub(crate) fn clear_and_remain_ready(&mut self) -> Result<()> {
         assert!(self.dirty);
-        // madvise the image range. This will throw away dirty pages,
-        // which are CoW-private pages on top of the initial heap
-        // image memfd.
-        unsafe {
-            rustix::io::madvise(
-                self.base as *mut c_void,
-                self.cur_size,
-                rustix::io::Advice::LinuxDontNeed,
-            )?;
+
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                // On Linux we can use `madvise` to reset the virtual memory
+                // back to its original state. This means back to all zeros for
+                // anonymous-backed pages and back to the original contents for
+                // CoW memory (the initial heap image). This has the precise
+                // semantics we want for reuse between instances, so it's all we
+                // need to do.
+                unsafe {
+                    rustix::io::madvise(
+                        self.base as *mut c_void,
+                        self.cur_size,
+                        rustix::io::Advice::LinuxDontNeed,
+                    )?;
+                }
+            } else {
+                // If we're not on Linux, however, then there's no generic
+                // platform way to reset memory back to its original state, so
+                // instead this is "feigned" by resetting memory back to
+                // entirely zeros with an anonymous backing.
+                //
+                // Additionally the previous image, if any, is dropped here
+                // since it's no longer applicable to this mapping.
+                self.reset_with_anon_memory()?;
+                self.image = None;
+            }
         }
 
         // mprotect the initial heap region beyond the initial heap size back to PROT_NONE.
-        self.set_protection(
-            self.initial_size..self.cur_size,
-            rustix::io::MprotectFlags::empty(),
-        )?;
+        self.set_protection(self.initial_size..self.cur_size, region::Protection::NONE)?;
         self.cur_size = self.initial_size;
         self.dirty = false;
         Ok(())
     }
 
-    fn set_protection(&self, range: Range<usize>, flags: rustix::io::MprotectFlags) -> Result<()> {
+    fn set_protection(&self, range: Range<usize>, flags: region::Protection) -> Result<()> {
         assert!(range.start <= range.end);
         assert!(range.end <= self.static_size);
         let mprotect_start = self.base.checked_add(range.start).unwrap();
         if range.len() > 0 {
             unsafe {
-                rustix::io::mprotect(mprotect_start as *mut _, range.len(), flags)?;
+                region::protect(mprotect_start as *mut _, range.len(), flags)?;
```

Comment:
Any particular reason this uses `region` while the other mmap-related bits use `rustix`? (The awkward `MprotectFlags` maybe? Or stemming from an attempt to get this working on Windows too?)

We actually have an interesting diversity of wrappers around `mprotect` in the codebase -- `region::protect` seems more common based on a cursory grep just now, while `rustix::io::mprotect` is used in the fiber library, here in memfd, and to set up a signal stack. So this is probably a larger question of "which syscall wrapper do we prefer". But IMHO at least within one unit of abstraction (the `MemFdSlot`) we should probably stick to one or else comment why we choose either one at each callsite...

Reply:
OK, yeah, that makes sense. There's a broader story here about how we are maybe wanting to use `rustix` for most system-interfacing things (@sunfishcode comment?) but that's definitely in tension with reusing existing crates that do what we want. Here given the use of rustix in surrounding code I think I'd tie-break toward using rustix but will defer to what @sunfishcode prefers.